### DISCUSSION TITLE
Allow beatmaps to enforce mods during specified sections

### DISCUSSION THREAD

**Discussion by felipemarins** [ppy/osu#28729] [Ideas]
This is something I thought as part of a proposal for a new beatmap format that I call **dynamic beatmaps** (certainly I wasn't the first one that thought about this, but I think it's a cool name). These beatmaps would change certain gameplay aspects dynamically as specified by the mapper (e.g., enabling mods during the middle of the map, that is what this discussion is about). While I would like to see this concept being developed even further (e.g., applying movement paths for circles or dynamic circle opacity), I think restricting it to mods would be a good start. So...

### What is this feature about?

To allow mappers to define certain sections where specific mods will be enabled. Possibly, also allowing to automate changes to these mods settings (e.g., to smoothly increase the approach rate using the Difficulty Adjust mod). I think it would also be nice that if the mappers could define a transition duration, as a fade-in or something, between the disabled and enabled mod state to not seem so abrupt for the player. If this overly complicates things, I guess a global transition duration could be implemented for a simpler implementation, but I think some mappers could want more control about this.

### Why implement this?

I think it would allow mappers to create more interesting experience for players and better represent the song in some situations. For example, a mapper could enforce the Flashlight mod for a sudden calmer section in the middle of a intense beatmap.

### How about stable compatibility?

I imagine that this could be implemented using a new file for the usage of this feature (which would store the timestamps for each mod usage) separated from the current .osu beatmap file. So stable would ignore this new file and just load the .osu file (that would contain the actual map).

### But what if I don't want to play with the enforced mods on lazer?

I expect that something like a new conversion mod named "Static" (as the opposite of "dynamic") could be created in order to tell lazer to ignore the new file.

### How would it work in conjunction to mods enabled by users?

I guess that mods in the "incompatible mods" list that each mod already have could be marked as incompatible in the mod select screen. If this still can lead to a lot of unexpected behaviour, I guess that all mods can be marked as incompatible if you don't use the "Static" mod (except for the automation mods).

### Some possible concerns

I guess that this can be a very complicated feature to implement with a lot of things to consider, such as transitions for each mod, SR algorithm, ranking criteria, etc. Overall, I think that this could turn into a programming hell for pp devs, so maybe the "dynamic" versions of these beatmaps could just live in a similar situation to loved beatmaps in which it doesn't give pp but it still has a leaderboard.

**Comment by shiumano** [comment]
Well, until now, that is what we have been able to achieve with the storyboard on the map.
Of course, it doesn't change anything but the look, so it doesn't do much more than that.

**Reply by felipemarins** [reply]
Yeah, I agree that storyboards add some dynamic to beatmaps and that they can be pretty interesting, but, yes, it's mostly a background thing and most players disable them. I think that with this ability to enforce mods, players would consider to play with these "dynamic" features, specially if the dynamic version had its own separate leaderboard instead of it being shared with the static version.

**Comment by Tenexxt** [comment]
Imo a great idea, it will be a hell to execute because programming it and thinking of how to go with the score multiplier (since in some parts there are mods, Score multi has to be applied, so it could be confusing ASF on ScoreV2). Also a great thing to add would be able to change the maps stats in the middle of the map to a certain degree, like make the HP drop faster, AC is faster etc.

**Reply by felipemarins** [reply]
> how to go with the score multiplier

For a simpler implementation, I think that all parts of the map could be treated the same in the scoring algorithm and the max score be 1M in the standardised score. If you want to consider the mod multiplier for each section, I think that some proportional multiplier to the judgement values could be applied. For example, a "Great" judgement in a section with HD applied could value 318 instead of 300. So, in a hypothetical map where there are just 2 circles, one in a section with NM and another in a section with HD, missing the HD circle would give you a total accuracy score of approximately 485 436. But, yes, I agree that this could be confusing.

> Also a great thing to add would be able to change the maps stats in the middle of the map to a certain degree, like make the HP drop faster, AC is faster etc.

I think this could be achieved by the usage of the Difficulty Adjust mod as I mentioned in the text.

**Comment by shiumano** [comment]
> I guess that some mods could be marked as incompatible in the mod select screen. If this leads to a lot of unexpected behaviour, I guess that all mods can be marked as incompatible if you don't use the "Static" mod.

Well, what about TouchDevice?
Not that it matters except for Autopilot, but I think it needs to be considered.

**Reply by felipemarins** [reply]
I guess I should have explained this part better, my bad. When I was talking about mods that could cause unexpected behaviour, I was referring to mods that change some gameplay aspect in a way that could interfere with the mods used in the map. I think that Touch Device and all automation mods can be allowed without any issues.
I hope that it would be possible to enable any mod compatible (as in "it's not in the incompatible mods list that each mod already have") with the mods enforced by the beatmap and I just added the last part in case it's too much trouble.

Edit: I edited the text to explain it better.
